### PR TITLE
[BREAKING] Disallow conversions between bytesX and uintY of different size.

### LINKED ISSUE #3786: Disallow conversion between bytesX and uintX of different sizes
Conversions between bytes and integer types always confuse people. It is rather straightforward to convert between bytes and uint types of the same size (nothing happens on the high-level side, alignment change on the low-level side). Size-changes inside the same type are also not too difficult (bytesXX is left-aligned, i.e. padding or cutting on the right, uintXX is right-aligned, i.e. padding or cutting on the left).

If both happens at the same time, users basically have to guess what happens first. If we just disallow such conversions, users have to explicitly state the order:

```
uint16 x = 0xffff;
bytes32(uint256(x)); // pad on the left
bytes32(bytes2(x)); // pad on the right
```

Inspiration: https://twitter.com/LefterisJP/status/977515366900555776

### PR DESCRIPTION
**Pull request by ekpyron**
Closes #3786.

This PR contains a simple, but breaking solution.
As mentioned in the issue, it may be possible to use a pragma 050 solution instead, but this would depend on #3879.

### CODE REVIEW

**Comment by ekpyron** [comment]
The second commit in this PR allows conversions from zero to bytesX (e.g. ``bytes32(0)``).
Since padding is irrelevant for this conversion I think it makes sense to make an exception here.

Do you agree?

In general conversions from rational numbers to bytesX types have the same problem as integer to bytesX conversions (is ``bytes32(0x1234)`` left or right aligned?), so I would disallow those as well (as is already the case in the PR).

**Comment by ekpyron** [comment]
Related to the last comment: #3942

**Comment by ekpyron** [comment]
Damn - I just noticed that this is actually a duplicate of #3899. I hadn't seen that, since the issue was still unassigned (not sure whether #3899 already referenced the issue, when I started - I might just have missed that).

**Comment by ekpyron** [comment]
I replaced the commits in this PR by a simplified version that only disallows conversions between bytesX and uintY of different size, but does not yet change the conversion between literals and bytesX types. I think it's best to handle the latter in a separate PR.

File: libsolidity/ast/Types.cpp [hunk 1]
```diff
@@ -884,7 +884,10 @@ bool RationalNumberType::isImplicitlyConvertibleTo(Type const& _convertTo) const
 bool RationalNumberType::isExplicitlyConvertibleTo(Type const& _convertTo) const
 {
 	TypePointer mobType = mobileType();
-	return mobType && mobType->isExplicitlyConvertibleTo(_convertTo);
+	return
+		(mobType && mobType->isExplicitlyConvertibleTo(_convertTo)) ||
+		(!isFractional() && _convertTo.category() == Category::FixedBytes)
```
Comment (ekpyron) [inline libsolidity/ast/Types.cpp:889]: This still allows all conversion between integer literals and bytesX types. This should be restricted in a separate PR and is tracked in #3942.

File: libsolidity/ast/Types.cpp [hunk 2]
```diff
@@ -476,7 +476,7 @@ bool IntegerType::isExplicitlyConvertibleTo(Type const& _convertTo) const
 	return _convertTo.category() == category() ||
 		_convertTo.category() == Category::Contract ||
 		_convertTo.category() == Category::Enum ||
-		_convertTo.category() == Category::FixedBytes ||
+		(_convertTo.category() == Category::FixedBytes && storageBytes() == _convertTo.storageBytes()) ||
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: `storageBytes` is a deduced property and might change in the future. IMHO it is better to use the native functions and mulitply by 8.

File: libsolidity/ast/Types.cpp [hunk 3]
```diff
@@ -1281,7 +1284,7 @@ bool FixedBytesType::isImplicitlyConvertibleTo(Type const& _convertTo) const
 
 bool FixedBytesType::isExplicitlyConvertibleTo(Type const& _convertTo) const
 {
-	return _convertTo.category() == Category::Integer ||
+	return (_convertTo.category() == Category::Integer && storageBytes() == _convertTo.storageBytes()) ||
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Same here.

**Comment by chriseth** [comment]
Fine to merge apart from the two issues above.

**Review by chriseth — APPROVED** [review]
